src/frontend/Typechecker.ml

```diff
@@ -0,0 +1,1549 @@
+  ; "short"; "signed"; "sizeof"; "static"; "static_assert"; "static_cast"
+  ; "struct"; "switch"; "template"; "this"; "thread_local"; "throw"; "true"
+  ; "try"; "typedef"; "typeid"; "typename"; "union"; "unsigned"; "using"
+  ; "virtual"; "void"; "volatile"; "wchar_t"; "while"; "xor"; "xor_eq"
+  ; "functions"; "data"; "parameters"; "model"; "return"; "if"; "else"; "while"
+  ; "for"; "in"; "break"; "continue"; "void"; "int"; "real"; "complex"
+  ; "vector"; "row_vector"; "matrix"; "ordered"; "positive_ordered"; "simplex"
+  ; "unit_vector"; "cholesky_factor_corr"; "cholesky_factor_cov"; "corr_matrix"
+  ; "cov_matrix"; "print"; "reject"; "target"; "get_lp"; "profile" ]
+
+let verify_identifier id : unit =
+  if id.name = !model_name then
+    Semantic_error.ident_is_model_name id.id_loc id.name |> error
+  else if
+    String.is_suffix id.name ~suffix:"__"
+    || List.mem reserved_keywords id.name ~equal:String.equal
+  then Semantic_error.ident_is_keyword id.id_loc id.name |> error
+  else ()
+
+let verify_name_fresh_var tenv id =
+  if Utils.is_unnormalized_distribution id.name then
+    Semantic_error.ident_has_unnormalized_suffix id.id_loc id.name |> error
+  else if
+    Env.mem tenv id.name
+    && not (Stan_math_signatures.is_stan_math_function_name id.name)
+  then Semantic_error.ident_in_use id.id_loc id.name |> error
+  else ()
+
+let verify_name_fresh_udf tenv id =
+  if
+    Stan_math_signatures.is_stan_math_function_name id.name
+    (* variadic functions are currently not in math sigs *)
+    || Stan_math_signatures.is_reduce_sum_fn id.name
+    || Stan_math_signatures.is_variadic_ode_fn id.name
+  then Semantic_error.ident_is_stanmath_name id.id_loc id.name |> error
+  else if Utils.is_unnormalized_distribution id.name then
+    Semantic_error.udf_is_unnormalized_fn id.id_loc id.name |> error
+  else if Env.mem tenv id.name then
+    (* adapt for overloading later *)
+    Semantic_error.ident_in_use id.id_loc id.name |> error
+  else ()
+
+(** Checks that a variable/function name:
+  - a function/identifier does not have the _lupdf/_lupmf suffix
+  - is not already in use (for now)
+*)
+let verify_name_fresh tenv id ~is_udf =
+  let f =
+    if is_udf then verify_name_fresh_udf tenv else verify_name_fresh_var tenv
+  in
+  List.iter ~f (probability_distribution_name_variants id)
+
+let is_of_compatible_return_type rt1 srt2 =
+  UnsizedType.(
+    match (rt1, srt2) with
+    | Void, NoReturnType
+     |Void, Incomplete Void
+     |Void, Complete Void
+     |Void, AnyReturnType ->
+        true
+    | ReturnType UReal, Complete (ReturnType UInt) -> true
+    | ReturnType UComplex, Complete (ReturnType UReal) -> true
+    | ReturnType UComplex, Complete (ReturnType UInt) -> true
+    | ReturnType rt1, Complete (ReturnType rt2) -> rt1 = rt2
+    | ReturnType _, AnyReturnType -> true
+    | _ -> false)
+
+(* -- Expressions ------------------------------------------------- *)
+let check_ternary_if loc pe te fe =
```

Comment:
These names are also a holdover. If I was guessing it would be something like 'predicate expression' 'true expression' 'false expression'.

The type hints aren't great in VSCode because I don't believe Merlin subsitutes in definitions of types unless they exactly match(`Ast.typed_expression = (typed_expr_meta, fun_kind) expr_with`. This function doesn't care what the second piece of that tuple is, so the type inference goes broader than our type definition)

In general, I tried to stick to expression variable names starting with `t` being typed expressions. This is one place that didn't happen, and I'm sure there are others.